python/sglang/srt/managers/utils.py

```
@@ -42,3 +47,51 @@ def validate_input_length(
             return error_msg
 
     return None
+
+
+# global expert distribution recording
+class ExpertDistributionRecorder:
+    # This class is a singleton class
+    def __new__(cls):
+        if not hasattr(cls, "instance"):
+            cls.instance = super(ExpertDistributionRecorder, cls).__new__(cls)
+        return cls.instance
+
+    def __init__(self):
+        # the length of the list is the number of tokens
+        # the length of the tuple is topk's k value
+        self._expert_distribution_record: List[Tuple[int]] = []
+        self._record = False
+
+    def record_new_token(self, topk_ids):
+        if not self._record:
+            return
+        topk_ids_list = topk_ids.to("cpu", non_blocking=True).numpy().tolist()
+        torch.cuda.synchronize()
+        for i in topk_ids_list:
+            self._expert_distribution_record.append(tuple(i))
+
+    def reset(self):
+        """Reset the expert distribution recorder."""
+        self._expert_distribution_record.clear()
+
+    def start_record(self):
+        """Start recording the expert distribution. Reset the recorder and set the recording flag to True."""
+        self.reset()
+        self._record = True
+
+    def stop_record(self):
+        """Stop recording the expert distribution. Set the recording flag to False."""
+        self._record = False
+
+    def dump_record(self):
+        """Dump the expert distribution record to a file. Reset the recorder after dumping."""
+        results = defaultdict(int)
+        for token_record in self._expert_distribution_record:
+            for expert_idx in token_record:
+                results[expert_idx] += 1
+        with open(f"expert_distribution_{time.time()}.csv", "w") as fd:
+            fd.write("expert_id,count\n")
```

My expected output format is `layer_id,expert_id,count` as different layers' experts are independent. But the current output only stores `expert_id` and `count`. Could you please explain it?

---

@ch-wan Previously I included layer_id here, but the layer_id information can only be retrieved from the model files so I removed it.  I added the layer id recording in `deepseek_v2.py` previously.  If we want to add it back, we'll need to add it to all the current MoE models, and ask developers to add to all the future MoE models.  I haven't thought of a better way to address the issue so I removed it.  If you think there is a better way, or it's fine to add to the model files, I can add it back.  Here is my original code for recording layer id: https://github.com/sgl-project/sglang/pull/4435/commits/50c0e0847ab835df97add252691f90b4fcd57845#diff-5b9e34dd492bd8a14702a18b594721091092276fad1cf8736fba6ef1f33c1b04R1167

---

I understand it. Mixing experts' workload distribution across layers has limited effective information. Your previous implementation looks good to me. We can support other models later. In `record_new_token`, we can check if `layer_id` is set or not. If not, we can throw an error and let users know that they are probably using a model that does not support recording experts workload.

---

Added the layer id recording back.  Please check.  Thanks!